tests/functional/api/test_errors.py

```diff
@@ -0,0 +1,195 @@
+        res = app.post_json(
+            "/api/groups", group, headers=auth_client_header, expect_errors=True
+        )
+
+        assert res.status_code == 403
+
+
+@pytest.mark.functional
+class Test404ErrorsUsingCreateGroup(object):
+    def test_it_returns_http_404_if_endpoint_nonexistent(self, app):
+        res = app.get("/api/not_a_thing", expect_errors=True)
+
+        assert res.status_code == 404
+
+    def test_it_returns_formatted_reason_if_endpoint_nonexistent(self, app):
+        res = app.get("/api/not_a_thing", expect_errors=True)
+
+        assert (
+            res.json["reason"]
+            == "Either the resource you requested doesn't exist, or you are not currently authorized to see it."
+        )
+
+
+@pytest.mark.functional
+class Test409ErrorsUsingCreateGroup(object):
+    def test_it_returns_HTTP_Conflict_if_groupid_is_duplicate(
+        self, app, auth_client_header, third_party_user
+    ):
+        headers = auth_client_header
+        headers[native_str("X-Forwarded-User")] = native_str(third_party_user.userid)
+        group = {"name": "My Group", "groupid": "group:333vcdfkj~@thirdparty.com"}
+
+        res = app.post_json("/api/groups", group, headers=headers)
+        res = app.post_json("/api/groups", group, headers=headers, expect_errors=True)
+
+        assert res.status_code == 409
+
+    def test_it_returns_formatted_reason_if_group_is_duplicate(
+        self, app, auth_client_header, third_party_user
+    ):
+        headers = auth_client_header
+        headers[native_str("X-Forwarded-User")] = native_str(third_party_user.userid)
+        group = {"name": "My Group", "groupid": "group:333vcdfkj~@thirdparty.com"}
+
+        res = app.post_json("/api/groups", group, headers=headers)
+        res = app.post_json("/api/groups", group, headers=headers, expect_errors=True)
+
+        assert (
+            res.json["reason"]
+            == "group with groupid 'group:333vcdfkj~@thirdparty.com' already exists"
+        )
+
+
+@pytest.fixture
+def third_party_user(db_session, factories):
+    user = factories.User(authority="thirdparty.com")
+    db_session.commit()
+    return user
+
+
+@pytest.fixture
+def auth_client(db_session, factories):
+    auth_client = factories.ConfidentialAuthClient(
+        authority="thirdparty.com", grant_type=GrantType.client_credentials
+    )
+    db_session.commit()
+    return auth_client
+
+
+@pytest.fixture
+def auth_client_header(auth_client):
+    user_pass = "{client_id}:{secret}".format(
+        client_id=auth_client.id, secret=auth_client.secret
+    )
+    encoded = base64.standard_b64encode(user_pass.encode("utf-8"))
+    return {
+        native_str("Authorization"): native_str(
+            "Basic {creds}".format(creds=encoded.decode("ascii"))
+        )
+    }
+
+
+@pytest.fixture
+def user_with_token(db_session, factories):
+    user = factories.User()
+    token = factories.DeveloperToken(userid=user.userid)
+    db_session.add(token)
```

Comment:
I don't think `add()` is necessary here, creating a model object with a factory automatically adds it to the DB session